src/Domain/Creators/ExpandOverBlocks.cpp

```diff
@@ -4,39 +4,97 @@
 #include "Domain/Creators/ExpandOverBlocks.hpp"
 
 #include <array>
+#include <boost/algorithm/string/join.hpp>
 #include <cstddef>
 #include <exception>
 #include <string>
 #include <vector>
 
+#include "Utilities/ErrorHandling/Assert.hpp"
 #include "Utilities/GenerateInstantiations.hpp"
 #include "Utilities/MakeArray.hpp"
 
 namespace domain {
 
+template <typename T, size_t Dim>
+ExpandOverBlocks<T, Dim>::ExpandOverBlocks(size_t num_blocks) noexcept
+    : num_blocks_(num_blocks) {}
+
+template <typename T, size_t Dim>
+ExpandOverBlocks<T, Dim>::ExpandOverBlocks(
+    std::vector<std::string> block_names,
+    std::unordered_map<std::string, std::unordered_set<std::string>>
+        block_groups) noexcept
+    : num_blocks_(block_names.size()),
+      block_names_(std::move(block_names)),
+      block_groups_(std::move(block_groups)) {}
+
 template <typename T, size_t Dim>
 std::vector<std::array<T, Dim>> ExpandOverBlocks<T, Dim>::operator()(
     const T& value) const {
-  return {num_blocks, make_array<Dim>(value)};
+  return {num_blocks_, make_array<Dim>(value)};
 }
 
 template <typename T, size_t Dim>
 std::vector<std::array<T, Dim>> ExpandOverBlocks<T, Dim>::operator()(
     const std::array<T, Dim>& value) const {
-  return {num_blocks, value};
+  return {num_blocks_, value};
 }
 
 template <typename T, size_t Dim>
 std::vector<std::array<T, Dim>> ExpandOverBlocks<T, Dim>::operator()(
     const std::vector<std::array<T, Dim>> value) const {
-  if (value.size() != num_blocks) {
+  if (value.size() != num_blocks_) {
     throw std::length_error{"You supplied " + std::to_string(value.size()) +
                             " values, but the domain creator has " +
-                            std::to_string(num_blocks) + " blocks."};
+                            std::to_string(num_blocks_) + " blocks."};
   }
   return value;
 }
 
+template <typename T, size_t Dim>
+std::vector<std::array<T, Dim>> ExpandOverBlocks<T, Dim>::operator()(
+    const std::unordered_map<std::string, std::array<T, Dim>>& value) const {
+  ASSERT(num_blocks_ == block_names_.size(),
```

Comment:
should this also be throwing or can this error never be triggered from an input file?

Reply:
no this is a developer error, they should pass block names to the constructor when they invoke the object with a map-over-block-names in the `std::variant`.